server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVChoices.java

```diff
@@ -21,5 +21,234 @@
+          break;
+        }
+      }
+
+      return new IRVChoices(mutableChoices);
+    }
+
+    /* Applies Rule 26.7.3 from
+     * https://www.sos.state.co.us/pubs/rule_making/CurrentRules/8CCR1505-1/Rule26.pdf
+     * This removes duplicated preferences for the same candidates.
+     * Returns new updated ballot.
+     */
+    public IRVChoices ApplyRule3() {
+
+      assert IsSorted(choices) : "Error: Trying to apply Rule 3 to unsorted preferences.";
+
+      HashSet<String> candidates = new HashSet<>();
+      List<IRVPreference> mutableChoices = new ArrayList<IRVPreference>();
+
+      for (IRVPreference choice : choices) {
+        if (!candidates.contains(choice.candidateName)) {
+          // This candidate hasn't been mentioned before. Add it to both the vote and the set of mentioned candidates.
+          mutableChoices.add(choice);
+          candidates.add(choice.candidateName);
+        }
+      }
+
+      return new IRVChoices(mutableChoices);
+    }
+
+    /* Returns this IRV vote as a sorted list of candidate names, starting with the first preference
+     * and ending with the lowest.
+     * Throws an exception if called on an invalid vote.
+     */
+    public List<String> AsSortedList() throws IRVParsingException {
+      // Neither of these calls should ever be made.
+      if ( !IsValid() ) {
+        throw new IRVParsingException("Attempt to call AsSortedList on invalid vote: "+choices);
+      }
+      if (! IsSorted(choices)) {
+        throw new IRVParsingException("Attempt to call AsSortedList on unsorted choices.");
+      }
+
+      return choices.stream().map(p -> p.candidateName).collect(Collectors.toUnmodifiableList());
+    }
+
+    /** Print as a human-readable string.
+     *  This will include explicit preferences in parentheses.
+     * Intended for votes that may not be valid,
+     * e.g. "Alice(1),Bob(1),Chuan(3)".
+     */
+    public String toString() {
+      return choices.stream().map(IRVPreference::toString).collect(Collectors.joining(","));
+    }
+
+  /**
+   * Parse a string (from the database) into a list of IRV preferences. This checks each individual
+   * preference for validity (a nonempty string, followed by a positive integer in parentheses) but
+   * does _not_ check whether the vote is valid - repeated and skipped preferences are allowed.
+   * @param sanitizedChoices a string describing the IRV vote, in the form "name1(p1),name2(p2),..."
+   * @return the same information as a list of IRVPreference objects.
+   * @throws IRVParsingException if any of the strings cannot be parsed as an IRVPreference.
+   */
+  private static List<IRVPreference> parseChoices(String sanitizedChoices) throws IRVParsingException {
+    ArrayList<IRVPreference> mutableChoices = new ArrayList<>();
+
+    String[] preferences = sanitizedChoices.trim().split(",");
+
+    for (String preference : preferences) {
+      mutableChoices.add(new IRVPreference(preference));
+    }
+
+    return mutableChoices;
+  }
+
```

Comment:
1. It should just use its field choices rather than taking an argument.
2. Yes, we really don't need it except as a defence against future mistakes in code editing. This part of the code is so fiddly, and so critical, that I felt the extra layer of defence was justified, rather than just writing a comment "this function assumes that the ranks are sorted so don't mess with that..." but I could easily be persuaded that that's overkill. I've left it in for now; happily persuaded to take it out.